src/decorator_trait.jl

```diff
@@ -0,0 +1,915 @@
+end
+
+function vector_transport_direction(
+    M::AbstractDecoratorManifold,
+    p,
+    X,
+    d,
+    m::AbstractVectorTransportMethod = default_vector_transport_method(M),
+)
+    return vector_transport_direction(trait(M, p, X, d, m), M, p, X, d, m)
+end
+function vector_transport_direction(
+    t::NestedTrait,
+    M::AbstractManifold,
+    p,
+    X,
+    d,
+    m::AbstractVectorTransportMethod = default_vector_transport_method(M),
+)
+    return vector_transport_direction(next_trait(t), M, p, X, d, m)
+end
+function vector_transport_direction(
+    ::EmptyTrait,
+    M::AbstractManifold,
+    p,
+    X,
+    d,
+    m::AbstractVectorTransportMethod = default_vector_transport_method(M),
+)
+    return invoke(
+        vector_transport_direction,
+        Tuple{AbstractManifold,typeof(p),typeof(X),typeof(d),typeof(m)},
+        M,
+        p,
+        X,
+        d,
+        m,
+    )
+end
+function vector_transport_direction(
+    ::NestedTrait{IsEmbeddedSubmanifoldManifold},
+    M::AbstractDecoratorManifold,
+    p,
+    X,
+    d,
+    m::AbstractVectorTransportMethod = default_vector_transport_method(M),
+)
+    return vector_transport_direction(get_embedding(M), p, X, d, m)
+end
+
+function vector_transport_direction!(
+    M::AbstractDecoratorManifold,
+    Y,
+    p,
+    X,
+    d,
+    m::AbstractVectorTransportMethod = default_vector_transport_method(M),
+)
+    return vector_transport_direction!(trait(M, Y, p, X, d, m), M, Y, p, X, d, m)
+end
+function vector_transport_direction!(
+    t::NestedTrait,
+    M::AbstractManifold,
+    Y,
+    p,
+    X,
+    d,
+    m::AbstractVectorTransportMethod = default_vector_transport_method(M),
+)
+    return vector_transport_direction!(next_trait(t), M, Y, p, X, d, m)
+end
+function vector_transport_direction!(
+    ::EmptyTrait,
+    M::AbstractManifold,
+    Y,
+    p,
+    X,
+    d,
+    m::AbstractVectorTransportMethod = default_vector_transport_method(M),
+)
+    return invoke(
+        vector_transport_direction!,
+        Tuple{AbstractManifold,typeof(Y),typeof(p),typeof(X),typeof(d),typeof(m)},
+        M,
+        Y,
+        p,
+        X,
+        d,
+        m,
+    )
+end
+function vector_transport_direction!(
+    ::NestedTrait{IsEmbeddedSubmanifoldManifold},
+    M::AbstractDecoratorManifold,
+    Y,
+    p,
+    X,
+    d,
+    m::AbstractVectorTransportMethod = default_vector_transport_method(M),
+)
+    return vector_transport_direction!(get_embedding(M), Y, p, X, d, m)
+end
+
+function vector_transport_to(
+    M::AbstractDecoratorManifold,
+    p,
+    X,
+    q,
+    m::AbstractVectorTransportMethod = default_vector_transport_method(M),
+)
+    return vector_transport_to(trait(M, p, X, q, m), M, p, X, q, m)
+end
+function vector_transport_to(
+    t::NestedTrait,
+    M::AbstractManifold,
+    p,
+    X,
+    q,
+    m::AbstractVectorTransportMethod = default_vector_transport_method(M),
+)
+    return vector_transport_to(next_trait(t), M, p, X, q, m)
+end
```

Comment:
What do you think about generating such pairs of functions with a macro? Would it be worth it?

Reply:
sure, it would reduce the number of code lines here tremendously.
One would generate all 3
* Decorator case that puts the trait in first parameter
* nested that calls next
* empty that invokes AbstractManifold

That would be a neat macro indeed, now that we are sure we want to stay on this path.

Reply:
OK, I'll work on that macro then.

Reply:
I've put the first two things into a macro in my last commit. Regarding the empty callback, are you sure it would always just forward to `AbstractManifold`?

Reply:
I am quite sure but we could make that type an optional argument (which defaults to `AbstractManifold`) ?
Though I can not imagine a case where you want to invoke somewhere else to.

Reply:
OK, I'll try to think of something that would make sense as a part of a separate package.

Reply:
One could – for the separate package – define some `default_empty_fallback()` function that returns the default type in the optional argument if it is not given (i.e. we set that function to return `AbstracManifold`)?

Reply:
It can't reliably work like that unfortunately, though it could probably be stored in a variable. I'll try a variable.

Reply:
You‘re more experienced in these. We could later write getter/setter for said variable, sure. But for now a variable is a good choice I think.

Reply:
For now I'm going to put all methods in a single macro without customization. I may split it later when we decide to move traits to a separate package.

Reply:
Ok, cool - sparing the lines for the fallback is cool and we can customise it later.